### ISSUE TITLE
Remove `e3nn` version pin

### ISSUE THREAD

**Issue by Andrew-S-Rosen** [ACEsuit/mace#555]
This is a feature request to ultimately remove the version pin of `e3nn==0.4.4` in `setup.cfg` to allow for more flexible versions (if possible).

**Comment by ilyes319** [comment]
I tested it, and we seem to support the latest e3nn version. This was not an easy decision, as it could weaken code interoperability, especially given how poorly TorchScript behaves. In which context is this problematic for you?

**Comment by Andrew-S-Rosen** [comment]
In my personal opinion, interoperability should be dependent on the individual packages being used. So, if `mace` supports all versions of `e3nn` but Package B doesn't, then Package B should be the one with a specific restriction on `e3nn`. But I understand that reality is a bit messier than this.

For me, it is not a huge issue. I package several machine learning workflows in my code [quacc](https://github.com/Quantum-Accelerators/quacc), which includes but is not limited to MACE. I allow users to `pip install quacc[ml]` to install the ML dependencies, but because `mace` is pinned to a specific version of `e3nn`, and some other libraries require `e3nn>0.4.4` (I unfortunately don't remember which), they couldn't be installed together even though they are potentially compatible.

**Comment by corinwagen** [comment]
Agreed, this would be really nice. `e3nn==0.4.4` is now almost three years old, and newer packages like `fairchem` require `e3nn>=0.5.0`.

**Comment by ilyes319** [comment]
It is unlikely we will support it in the short term because it breaks backward compatibility. You are welcome to install 0.5 or later, the package will still work.

**Comment by Andrew-S-Rosen** [comment]
Ah, that's unfortunate. I don't know that I agree with the philosophy here, but I can understand the motivation. In what way does it break backwards compatability? Do energies change with the newer version of e3nn? If so, shouldn't `mace` want to use this presumably more accurate model architecture? If it is incompatible, semantic versioning with an appropriate CHANGELOG would hopefully address this. 

It is not practical to install 0.5 or later from a package management standpoint. I can't put e3nn>=0.5 in `pyproject.toml`, for instance, because that would conflict with `mace` if it is also in there.

**Comment by gabor1** [comment]
Can someone explain to me the issue? Surely all packages should allow the broadest possible (tested) set of versions to maximise flexibility within the constraint of correctness

**Comment by gabor1** [comment]
This does not preclude defining a more restricted subset of packages for vanilla installs whose goal is to simplify the installation process and minimise package conflicts. But this goal should not be conflated with the goal of allowing flexibility of someone wants it

**Comment by ilyes319** [comment]
e3nn changed the phase convention of the CG making the code not backward compatible and bunch of tests failing. There are ways to solve this, but it will be in the next release.

**Comment by Andrew-S-Rosen** [comment]
Ah okay, thanks for the clarity! I had assumed based on [your prior comment](https://github.com/ACEsuit/mace/issues/555#issuecomment-2305820396) that everything was still passing. Of course, failing tests should be a clear sign not to change the version yet.

**Comment by ilyes319** [comment]
The plan is to release 0.3.7 when all the bugs are fixed and then bump the version to 0.4.0 and update e3nn version to the latest, updating the tests. I would say a 2 months timeline.

**Comment by corinwagen** [comment]
Thanks for the update; makes total sense!

**Comment by bkmi** [comment]
I'm curious if there is progress on this. I imagine it's a rather large change actually. All models would either have to be retrained, or the basis transformation would have to be applied regularly throughout the network.

Any thoughts?

**Comment by vsimkus** [comment]
I'll add that because of the dependency on `e3nn==0.4.4` mace currently cannot be used with `torch>=2.6.0`. The cause is [this line](https://github.com/e3nn/e3nn/blob/5c0d2fdcc719190e40a887ddb5bd5790e196ba5f/e3nn/o3/_wigner.py#L10C1-L10C99) in `e3nn==0.4.4`, which breaks with `torch>=2.6.0` due to a vulnerability patch that [breaks backwards compatibility](https://dev-discuss.pytorch.org/t/bc-breaking-change-torch-load-is-being-flipped-to-use-weights-only-true-by-default-in-the-nightlies-after-137602/2573) of `torch.load`. This is addressed in `e3nn>=0.5.0`.

**Comment by vsimkus** [comment]
Yes, it fails when trying to import `MACECalculator` due to `from e3nn import o3`. 

```
In [2]: import torch
   ...: print(torch.__version__)
   ...: import e3nn
   ...: print(e3nn.__version__)
   ...: import mace
   ...: print(mace.__version__)
   ...: from mace.calculators import MACECalculator
2.6.0
0.4.4
0.3.10
---------------------------------------------------------------------------
UnpicklingError                           Traceback (most recent call last)
```

**Comment by ilyes319** [comment]
that should work, it is in the CI. Can you try to train?

**Comment by vsimkus** [comment]
> I have added a patch in 3.10 and mace perfectly works with 2.6. Did you actually get an error from the latest main?

Ah yes, sorry your previous comment confused me. It works on `main` but the patch is not in `0.3.10`. Thanks!

Looking forward to `0.4.0` where this is properly addressed allowing newer versions of `e3nn`!

**Comment by Andrew-S-Rosen** [comment]
Looks like this is still pinned in `main`? Not sure if I misunderstood.

**Comment by RichardJBlack** [comment]
If I understand this correctly, mace requires e3nn==0.4.4 which in turn requires torch < 2.6.0.  But torch must be at least 2.8.0 for security vulnerability reasons.  Hence mace is prohibited/unusable.  This issue is open 18 months now.  Should we assume that mace is abandoned project if 18 months pass without addressing this security/compliance issue.

**Comment by ilyes319** [comment]
Dear @RichardJBlack,
 1. MACE is compatible with PyTorch 2.8. I am using it, and so are most people I know. 
 2. Where is the generic statement « Torch must be at least 2.8.0 for security vulnerability reasons." coming from?
 3. MACE is an open-source project, and we do our best to communicate in a respectful and non-aggressive way. I would kindly request that you do the same before posting here. Thank you.

**Comment by vsimkus** [comment]
> Where is the generic statement « Torch must be at least 2.8.0 for security vulnerability reasons." coming from?

I believe they're referring to this: https://dev-discuss.pytorch.org/t/bc-breaking-change-torch-load-is-being-flipped-to-use-weights-only-true-by-default-in-the-nightlies-after-137602/2573.

**Comment by ilyes319** [comment]
Sure but that was changed at pytorch 2.6 not 2.8


________________________________

**Comment by ananth1996** [comment]
> Agreed, this would be really nice. `e3nn==0.4.4` is now almost three years old, and newer packages like `fairchem` require `e3nn>=0.5.0`.

Facing a similar issue in our project where we want to have `fairchem-core` and `mace`, which is causing problems with our package manager `uv`. The only way forward seems to be finding a workaround to install both requirements.

Is there any update/progress on this issue?

**Comment by ilyes319** [comment]
@ananth1996 Note that if you use the uma/esen models, you can safely downgrade back to 0.4.4 as they don't rely on e3nn. This is how i do it.

**Comment by ananth1996** [comment]
> [@ananth1996](https://github.com/ananth1996) Note that if you use the uma/esen models, you can safely downgrade back to 0.4.4 as they don't rely on e3nn. This is how i do it.

Hi @ilyes319 , so I took your advice and added the following to my `pyproject.toml` ( I am using `uv`):
```
[tool.uv]
override-dependencies = ["e3nn==0.4.4"]
```
This pins the `e3nn` to the older version, and then I can install `mace-torch` using `uv add mace-torch`. 
This works, and I can run the example `mace` scripts.

However, this causes my `fairchem-core` to break. When I try to load an uma model in `fairchem` following their [example](https://github.com/facebookresearch/fairchem?tab=readme-ov-file#run-molecular-md) I get the following error:
```
UnpicklingError: Weights only load failed. This file can still be loaded, to do so you have two options, do those steps only if you trust the source of the checkpoint. 
	(1) In PyTorch 2.6, we changed the default value of the `weights_only` argument in `torch.load` from `False` to `True`. Re-running `torch.load` with `weights_only` set to `False` will likely succeed, but it can result in arbitrary code execution. Do it only if you got the file from a trusted source.
	(2) Alternatively, to load with `weights_only=True` please check the recommended steps in the following error message.
	WeightsUnpickler error: Unsupported global: GLOBAL slice was not an allowed global by default. Please use `torch.serialization.add_safe_globals([slice])` or the `torch.serialization.safe_globals([slice])` context manager to allowlist this global if you trust this class/function.

Check the documentation of torch.load to learn more about types accepted by default with weights_only https://pytorch.org/docs/stable/generated/torch.load.html.
```

Upon following the error stack, this seems to be because I have `torch==2.8.0` and `e3nn==0.4.4`. 
The `fairchem` package loads `e3nn.o3` [here](https://github.com/facebookresearch/fairchem/blob/e8a8946a881a698396955b83b4980c136202d9eb/src/fairchem/core/models/uma/common/so3.py#L13). But `e3nn==0.4.4` has [this module-level `torch.load`](https://github.com/e3nn/e3nn/blob/5c0d2fdcc719190e40a887ddb5bd5790e196ba5f/e3nn/o3/_wigner.py#L10):
```
_Jd, _W3j_flat, _W3j_indices = torch.load(os.path.join(os.path.dirname(__file__), 'constants.pt'))
```
Which, by default, uses `weights_only=True` in PyTorch 2.8, causing issues when loading.


I can mitigate this by setting the `TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD` environment variable, which allows loading unsafe weights.

I think we are okay with this workaround for the time being, and we look forward to version `0.4.0` of `mace-torch` to remove these workarounds.

**Comment by ilyes319** [comment]
yep that is how i do it, you can also import mace before importing fairchem and it will work fine.

**Comment by cvjjm** [comment]
I recently ran into exactly the problem @ananth1996 and @ilyes319 talk about. Why does importing mace before fairchem work around the issue?

**Comment by Olivier7017** [comment]
Bumping this as the issue is affecting one of my project. 

I would like to reiterate that the security fix forced in PyTorch 2.6.0 around weights_only breaks e3nn==0.4.4. This means that any PyTorch version since January 29, 2025 is incompatible without extra workaround. It is becoming increasingly difficult to build a python environment without hitting this conflict, as most packages have by now moved to torch>=2.6.0. 

I noticed a new release was published on May 10 2026, but the e3nn pin remains unchanged. Is unpinning e3nn still being considered?

**Comment by ilyes319** [comment]
yes it is and we have a plan for that, it is going to happen in around August as part of the next major mace version release. More on the road map soon.

**Comment by IgnacioJPickering** [comment]
Is there any update on this? Depending on the old version of e3nn makes MACE incompatible with NequIP too https://github.com/mir-group/nequip/blob/27d9d2182da918ab7be0017d8300e53278f5e00e/pyproject.toml#L22
